src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultDocumentDivergenceChecker.cs

```diff
@@ -0,0 +1,304 @@
+            {
+                return false;
+            }
+
+            if (newProperty.Modifiers.Count != oldProperty.Modifiers.Count)
+            {
+                return false;
+            }
+
+            for (var i = 0; i < newProperty.Modifiers.Count; i++)
+            {
+                var newModifier = newProperty.Modifiers[i];
+                var oldModifier = oldProperty.Modifiers[i];
+                if (newModifier.Text != oldModifier.Text)
+                {
+                    return false;
+                }
+            }
+
+            if (newProperty.AttributeLists.Count != oldProperty.AttributeLists.Count)
+            {
+                return false;
+            }
+
+            for (var i = 0; i < newProperty.AttributeLists.Count; i++)
+            {
+                var newAttributes = newProperty.AttributeLists[i].Attributes;
+                var oldAttributes = oldProperty.AttributeLists[i].Attributes;
+
+                if (newAttributes.Count != oldAttributes.Count)
+                {
+                    return false;
+                }
+
+                for (var j = 0; j < newAttributes.Count; j++)
+                {
+                    var newAttribute = newAttributes[j];
+                    var oldAttribute = oldAttributes[j];
+
+                    if (!newAttribute.IsEquivalentTo(oldAttribute))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private static List<PropertyDeclarationSyntax> ExtractCSharpProperties(SourceText newText, IReadOnlyList<RazorDirectiveSyntax> codeBlocks)
+        {
+            var propertyList = new List<PropertyDeclarationSyntax>();
+            for (var i = 0; i < codeBlocks.Count; i++)
+            {
+                var bodyRange = codeBlocks[i].Body.Span;
+                var bodyTextSpan = TextSpan.FromBounds(bodyRange.Start, bodyRange.End);
+                var subText = newText.GetSubText(bodyTextSpan);
+                var parsedText = CSharpSyntaxTree.ParseText(subText);
+                var root = parsedText.GetRoot();
+                var childNodes = root.ChildNodes();
+                var properties = childNodes.Where(node => node.Kind() == CodeAnalysis.CSharp.SyntaxKind.PropertyDeclaration).OfType<PropertyDeclarationSyntax>();
+                propertyList.AddRange(properties);
+            }
+
+            return propertyList;
+        }
+
+        private bool TryGetSyntaxTreeAndSource(DocumentSnapshot document, out RazorSyntaxTree syntaxTree, out SourceText sourceText)
+        {
+            if (document.TryGetGeneratedOutput(out var codeDocument) &&
+                document.TryGetText(out sourceText))
+            {
+                syntaxTree = codeDocument.GetSyntaxTree();
+                return true;
+            }
+            else
+            {
+                if (!document.TryGetText(out sourceText))
```

Comment:
Kinda weird to have this and also the call on line 193, not 100% sure what my suggestio would be though.

Reply:
Hmm, we could pre-emptively check source text as a separate if in the beginning!